openmc/cmfd.py

```diff
@@ -2214,20 +2154,46 @@ def _neutron_balance(self):
         res = leakage + interactions - scattering - (1.0 / keff) * fission
 
         # Normalize res by flux and bank res
-        self._resnb = np.divide(res, self._flux, where=self._flux > 0)
+        self._resnb = np.divide(res, self._flux, where=self._flux > 0,
+                                out=np.zeros_like(self._flux))
 
         # Calculate RMS and record for this batch
         self._balance.append(np.sqrt(
             np.sum(np.multiply(self._resnb, self._resnb)) /
             (ng * num_accel)))
 
     def _precompute_array_indices(self):
-        """Computes the indices used to populate certain cross section arrays.
-        These indices are used in _compute_dtilde and _compute_dhat
+        """Initializes cross section arrays and computes the indices
+        used to populate dtilde and dhat
 
         """
         # Extract spatial indices
-        nx, ny, nz = self._indices[:3]
+        nx, ny, nz, ng = self._indices
+
+        # Allocate dimensions for each mesh cell
+        self._hxyz = np.zeros((nx, ny, nz, 3))
+        self._hxyz[:] = openmc.capi.meshes[self._mesh_id].width
+
+        # Allocate flux, cross sections and diffusion coefficient
+        self._flux = np.zeros((nx, ny, nz, ng))
+        self._totalxs = np.zeros((nx, ny, nz, ng))
+        self._p1scattxs = np.zeros((nx, ny, nz, ng))
+        self._scattxs = np.zeros((nx, ny, nz, ng, ng))  # Incoming, outgoing
+        self._nfissxs = np.zeros((nx, ny, nz, ng, ng))  # Incoming, outgoing
+        self._diffcof = np.zeros((nx, ny, nz, ng))
+
+        # Allocate dtilde and dhat
+        self._dtilde = np.zeros((nx, ny, nz, ng, 6))
+        self._dhat = np.zeros((nx, ny, nz, ng, 6))
+
+        # Set reference diffusion parameters
+        if list(self._ref_d):
```

Comment:
Since we force `_ref_d` to be a numpy array, why not just check
```suggestion
        if self._ref_d.size > 0:
```